### PR TITLE
Node selector support via constraints

### PR DESCRIPTION
**Pull request by alexellis**
## Description

Brings parity with Docker Swarm implementation.

We can now pin a function to a specific node such as:

```
constraints:
  - environment=production
```

*`faas-cli` YAML*

### Test plan:

First label a node using `kubectl`:

```
$ kubectl label node/node1 environment=true
```

* Next add a constraint to a function using the YAML example above and deploy. 

* You should see that the function is scheduled on the node. Now try removing the function and the constraint and schedule again. 

* Finally add a constraint that can never be matched and then reschedule the function, you should see it sitting at 0/1 replicas.

Next - do all of the above using the `--update=true --replace=false` flag to `faas-cli deploy`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by alexellis** [comment]
@johnmccabe @ericstoekl I'm looking to push this through. Would you be interested in getting involved testing? I know you may have multi-node clusters.

**Comment by ericstoekl** [comment]
Sure, I'll try this PR out. Can I add myself as a reviewer using Derek?

**Comment by rgee0** [comment]
@ericstoekl you should be able to start a review when commenting in the file changes

**Comment by alexellis** [comment]
@rayhero I think this could be useful for your image / opencv processing

**Comment by alexellis** [comment]
Eric would still appreciate a review & testing.

**Comment by johnmccabe** [comment]
Have spun this up locally, on a two node cluster however so I've tainted master.

My Kubernetes version is:
```
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.3", GitCommit:"f0efb3cb883751c5ffdbe6d515f3cb4fbe7b7acd", GitTreeState:"clean", BuildDate:"2017-11-08T18:27:48Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

Created the following label on `kube2`:
```
kubectl label node/kube2 environment=production
```
```
PS D:\Workspaces\go\src\github.com\openfaas\faas-netes> kubectl describe node kube2
Name:               kube2
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    environment=production
                    kubernetes.io/hostname=kube2
```

**Without constraint set**
Behaves as expected with pods spread across both cluster nodes.
```
faas-cli deploy --image johnmccabe/fn-echo --name echo --g
ateway http://kube1.johnmccabe.infra:31112
```
image

**With constraint set**
Behaves as expected with pods restricted to `kube2`
```
faas-cli.exe deploy --image johnmccabe/fn-env --name echo --gat
eway http://kube1.johnmccabe.net:31112 --constraint "environment=production"
```
Also tried with one of the default labels.
```
faas-cli deploy --image johnmccabe/fn-echo --name echo --g
ateway http://kube1.johnmccabe.infra:31112  --constraint "kubernetes.io/hostname=kube2"
```
image

File: handlers/deploy.go [hunk 1]
```diff
@@ -22,6 +23,12 @@ import (
 	"k8s.io/client-go/kubernetes"
 )
 
+// WatchdogPort for the OpenFaaS function watchdog
+const WatchdogPort = 8080
+
+// InitialReplicas how many replicas to start of creating for a function
+const InitialReplicas = 1
```
Comment (johnmccabe) [inline handlers/deploy.go]: Worth allowing the user to set the initial/min number of replicas when deploying a function?

Also does this actually need to be exported? its only used in this file.

File: handlers/deploy.go [hunk 2]
```diff
@@ -242,3 +254,21 @@ func buildEnvVars(request requests.CreateFunctionRequest) []v1.EnvVar {
...
 	return &i
 }
+
+func createSelector(constraints []string) map[string]string {
+	selector := make(map[string]string)
+
+	log.Println(constraints)
+	if len(constraints) > 0 {
+		for _, constraint := range constraints {
+			parts := strings.Split(constraint, "=")
```
Comment (johnmccabe) [inline handlers/deploy.go]: Sanity checked with a keyonly constraint here and it looked ok.
```
faas-cli.exe deploy --image johnmccabe/fn-env --name echo --gateway http://kube1.johnmccabe.infra:31112 --constraint "keyonlylabel="
```

File: handlers/deploy.go [hunk 3]
```diff
@@ -22,6 +23,12 @@ import (
 	"k8s.io/client-go/kubernetes"
 )
 
+// WatchdogPort for the OpenFaaS function watchdog
+const WatchdogPort = 8080
```
Comment (johnmccabe) [inline handlers/deploy.go]: Does this actually need to be exported? its only used in this file.

**Comment by johnmccabe** [comment]
@alexellis can this move to done in the project

**Comment by alexellis** [comment]
Do you have access?

File: handlers/deploy.go (hunk 1 above)
Comment (alexellis) [reply handlers/deploy.go]: Number 1 is out of scope for PR, but can come under the scope of the issue in the faas repo around setting max replicas.

**Comment by alexellis** [comment]
Thanks for taking time to test and validate

**Pull request merged**